examples/virtual_methods/base_button/mod.rs

```diff
@@ -0,0 +1,113 @@
+/// Public part of the BaseButton
+mod imp;
+
+use gtk::glib::{
+    self,
+    subclass::prelude::*,
+    translate::{from_glib_borrow, ToGlibPtr},
+    Error,
+};
+use gtk::prelude::*;
+use std::{future::Future, pin::Pin};
+
+glib::wrapper! {
+    pub struct BaseButton(ObjectSubclass<imp::BaseButton>)
+        @extends gtk::Widget, gtk::Button;
+}
+
+/// Public trait that implements our functions for everything that derives from BaseButton
+pub trait BaseButtonExt {
+    fn sync_method(&self);
+    fn async_method(&self) -> Pin<Box<dyn Future<Output = Result<(), Error>> + 'static>>;
+}
+
+/// We call into imp::BaseButton_$method_name for each function. These will retrieve the
+/// correct class (the base class for the BaseButton or the derived class for DerivedButton)
+/// and call the correct implementation of the function.
+impl<O: IsA<BaseButton>> BaseButtonExt for O {
+    fn sync_method(&self) {
+        unsafe { imp::BaseButton_sync_method(self.as_ref().to_glib_none().0) }
+    }
+
+    fn async_method(&self) -> Pin<Box<dyn Future<Output = Result<(), Error>> + 'static>> {
+        unsafe { imp::BaseButton_async_method(self.as_ref().to_glib_none().0) }
+    }
+}
+
+/// The BaseButtonImpl that each derived private struct has to implement. See derived_button/imp.rs for how
+/// to override functions.
+pub trait BaseButtonImpl: ObjectImpl + 'static {
+    fn sync_method(&self, obj: &BaseButton) {
+        self.parent_sync_method(obj)
+    }
+
+    fn async_method(
+        &self,
+        obj: &BaseButton,
+    ) -> Pin<Box<dyn Future<Output = Result<(), Error>> + 'static>> {
+        self.parent_async_method(obj)
+    }
+
+    fn parent_sync_method(&self, obj: &BaseButton) {
+        unsafe {
+            let data = Self::type_data();
+            let parent_class = data.as_ref().parent_class() as *mut imp::BaseButtonClass;
+            if let Some(ref f) = (*parent_class).sync_method {
+                f(obj.to_glib_none().0)
+            } else {
+                unimplemented!()
+            }
+        }
+    }
+
+    fn parent_async_method(
+        &self,
+        obj: &BaseButton,
+    ) -> Pin<Box<dyn Future<Output = Result<(), Error>> + 'static>> {
+        unsafe {
+            let data = Self::type_data();
+            let parent_class = data.as_ref().parent_class() as *mut imp::BaseButtonClass;
+            if let Some(ref f) = (*parent_class).async_method {
+                f(obj.to_glib_none().0)
+            } else {
+                unimplemented!()
+            }
+        }
+    }
+}
+
+/// Make the BaseButton subclassable
+unsafe impl<T: BaseButtonImpl> IsSubclassable<T> for BaseButton {
+    fn class_init(class: &mut glib::Class<Self>) {
+        <glib::Object as IsSubclassable<T>>::class_init(class);
```

Comment:
```suggestion
        <gtk::Button as IsSubclassable<T>>::class_init(class);
```
the same as for instance_init

Reply:
Hm, that makes rustc complain though:

```

error[E0308]: mismatched types
  --> examples/virtual_methods/base_button/mod.rs:76:56
   |
76 |         <gtk::Button as IsSubclassable<T>>::class_init(class);
   |                                                        ^^^^^ expected struct `gtk4::Button`, found struct `base_button::BaseButton`
   |
   = note: expected mutable reference `&mut Class<gtk4::Button>`
              found mutable reference `&mut Class<base_button::BaseButton>`
```

Not sure if I have to explicitly upcast at that point?